rasa_sdk/executor.py

```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
                 f"tracker, domain. Your function accepts only {len(valid_keys)} "
                 "parameters."
             )
+
+        if name in self.actions:
+            logger.info(f"Re-registered function for '{name}'.")
+        else:
+            logger.info(f"Registered function for '{name}'.")
+
         self.actions[name] = f
 
     def _import_submodules(
         self, package: Union[Text, types.ModuleType], recursive: bool = True
     ) -> None:
-        """ Import all submodules of a module, recursively, including
-        subpackages
+        """Import a module, or a package and its submodules recursively.
 
-        :param package: package (name or actual module)
-        :type package: str | module
-        :rtype: dict[str, types.ModuleType]
+        Args:
+            package: Package or module name, or an already loaded Python module.
+            recursive: If `True`, and `package` is a package, import all of its
+                sub-packages as well.
         """
         if isinstance(package, str):
-            package = importlib.import_module(package)
+            package = self._import_module(package)
 
         if not getattr(package, "__path__", None):
             return
 
-        results = {}
         for loader, name, is_pkg in pkgutil.walk_packages(package.__path__):
             full_name = package.__name__ + "." + name
-            results[full_name] = importlib.import_module(full_name)
+            self._import_module(full_name)
+
             if recursive and is_pkg:
                 self._import_submodules(full_name)
 
+    def _import_module(self, name: Text) -> types.ModuleType:
+        """Import a Python module. If possible, register the file where it came from.
+
+        Args:
+            name: The module's name using Python's module path syntax.
+
+        Returns:
+            The loaded module.
+        """
+        module = importlib.import_module(name)
+
+        module_file = getattr(module, "__file__", None)
+        if module_file:
+            # If the module we're importing is a namespace package (a package
+            # without __init__.py), then there's nothing to watch for the
+            # package itself.
+            timestamp = os.path.getmtime(module_file)
+            self._modules[module_file] = (timestamp, module)
+
+        return module
+
     def register_package(self, package: Union[Text, types.ModuleType]) -> None:
-        from rasa_sdk.interfaces import Action
+        """Register all the `Action` subclasses contained in a Python module or
+        package.
```

Comment:
🚀 Let's document that this function potentially kills everything in case loading the module goes wrong.